### PR TITLE
Stop `-O`/`-C opt-level` and `-g`/`-C debuginfo` conflicting

### LINKED ISSUE #7493: Can not override previous opt-levels in list of command flags
If one passes multiple opt-level flags to the compiler (for example -O, and --opt-level=3) then an error results. Instead, the previous setting should be overridden. This is bad because then makefiles don't work easily. For example, I had to use the command line `make RUSTFLAGS=--opt-level=3 DISABLE_OPTIMIZE=1` instead of the more obvious `make RUSTFLAGS=--opt-level=3` when compiling rust as optimized because of this error.

### LINKED ISSUE #32352: rustc should accept rustc -O -C opt-level=x
`rustc -O -O` is ok just like `rustc -C opt-level=1 -C opt-level=2`. `-O` is documented to be the same as `-C opt-level=2`, so there should be no conflict, just have the last argument take precedent.

`-g` vs `-C debuginfo=2` behave the same way.

```
$ rustc -O -C opt-level=2
error: -O and -C opt-level both provided

$ rustc -C opt-level=1 -C opt-level=2 -g -g foo.rs
# ok.
```

### PR DESCRIPTION
**Pull request by varkor**
Allow `-O` and `-C opt-level`, and `-g` and `-C debuginfo` to be specified simultaneously without conflict. In such cases, the rightmost provided option is chosen.

Fixes https://github.com/rust-lang/rust/issues/7493.
Fixes https://github.com/rust-lang/rust/issues/32352.

~Blocked on https://github.com/rust-lang-nursery/getopts/pull/79.~

r? @alexcrichton

### CODE REVIEW

File: src/test/run-make/override-aliased-flags/Makefile [hunk 1]
```diff
@@ -0,0 +1,22 @@
+-include ../../run-make-fulldeps/tools.mk
+
+# FIXME: it would be good to check that it's actually the rightmost flags
```
Comment (varkor) [inline src/test/run-make/override-aliased-flags/Makefile]: Ideas for checking this in the test are welcome!

File: src/librustc/session/config.rs [hunk 2]
```diff
@@ -2184,10 +2184,15 @@ pub fn build_session_options_and_crate_config(
         TargetTriple::from_triple(host_triple())
     };
     let opt_level = {
-        if matches.opt_present("O") {
-            if cg.opt_level.is_some() {
-                early_error(error_format, "-O and -C opt-level both provided");
+        let max_o = matches.opt_positions("O").into_iter().max();
```
Comment (nagisa) [inline src/librustc/session/config.rs]: NB: `max` is not what the common expectation will be for these flags. Instead users expect that the last flag takes precedence.

Also for `opt-level` it is not possible to establish absolute ordering between `s`, `z` and the numeric levels.

Comment (nagisa) [reply src/librustc/session/config.rs]: Ah never mind me, I read the code wrong and missed that this is taking the maximum of argument *position*.

**Comment by alexcrichton** [comment]
Looks reasonable to me! I think testing manually here is fine for now, I also don't know of a great way to test that the right setting is applied. 

Could this also include a small comment in the argument parsing about how it's finding the last option specified?

**Comment by varkor** [comment]
> Could this also include a small comment in the argument parsing about how it's finding the last option specified?

Done.

**Comment by alexcrichton** [comment]
r=me when this is ready to go

**Comment by Centril** [comment]
@bors rollup-

(don't rollup PRs that touch `.lock` files)

**Comment by jethrogb** [comment]
```
[00:59:29] ---- [run-make] run-make/override-aliased-flags stdout ----
[00:59:29] note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
[00:59:29] 
[00:59:29] error: make failed
[00:59:29] status: exit code: 2
[00:59:29] command: "make"
[00:59:29] stdout:
[00:59:29] ------------------------------------------
[00:59:29] # Test that `-O` and `-C opt-level` can be specified multiple times.
[00:59:29] # The rightmost flag will be used over any previous flags.
[00:59:29] LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/override-aliased-flags/override-aliased-flags:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc' --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/override-aliased-flags/override-aliased-flags -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/override-aliased-flags/override-aliased-flags  -Clinker=arm-none-eabi-gcc -O -O main.rs
[00:59:29] Makefile:8: recipe for target 'all' failed
[00:59:29] 
[00:59:29] ------------------------------------------
[00:59:29] stderr:
[00:59:29] ------------------------------------------
[00:59:29] error: linking with `arm-none-eabi-gcc` failed: exit code: 1
[00:59:29]   |
[00:59:29]   = note: "arm-none-eabi-gcc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/override-aliased-flags/override-aliased-flags/main.main.7rcbfp3g-cgu.0.rcgu.o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/override-aliased-flags/override-aliased-flags/main.main.7rcbfp3g-cgu.1.rcgu.o" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/override-aliased-flags/override-aliased-flags/main" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/override-aliased-flags/override-aliased-flags/main.4s37gsrti678ik8u.rcgu.o" "-Wl,--gc-sections" "-pie" "-Wl,-zrelro" "-Wl,-znow" "-Wl,-O1" "-nodefaultlibs" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/override-aliased-flags/override-aliased-flags" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,--start-group" "-Wl,-Bstatic" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-36963fe6a2961b22.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-bad302bb9c2be352.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libbacktrace_sys-068d4c647d225b00.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-dd618d3a61d2285d.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-08bd494bdfbe6ec5.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-f434e5c003b087ed.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-4ad3589c043e8a63.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-b93458722dc1ee7d.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-cbc983130e35415c.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-b986af5e424776e6.rlib" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-08dafd64384d2a0a.rlib" "-Wl,--end-group" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-bc40ea2e2336c57a.rlib" "-Wl,-Bdynamic" "-ldl" "-lrt" "-lpthread" "-lgcc_s" "-lc" "-lm" "-lrt" "-lpthread" "-lutil" "-lutil"
[00:59:29]   = note: arm-none-eabi-gcc: error: unrecognized command line option '-m64'
[00:59:29]           
[00:59:29] 
[00:59:29] error: aborting due to previous error
[00:59:29] 
[00:59:29] make: *** [all] Error 1
[00:59:29] 
[00:59:29] ------------------------------------------
```
https://travis-ci.com/rust-lang/rust/jobs/197566788 via #60519

**Comment by Zoxc** [comment]
These command line flags can conflict though. Why should we not emit an error?

**Comment by varkor** [comment]
@Zoxc: the original issue threads contain motivation. Flags in general can be specified multiple times and `-O` and `-C opt-level` are supposed to act like synonyms, so it's surprising that they conflict with each other.

**Comment by Zoxc** [comment]
I guess nonsensical things like `-C opt-level=2 -C opt-level=3` is unfortunately allowed already.

**Comment by petrochenkov** [comment]
It's not nonsensical, it's needed for hierarchical build systems.
The top level project sets default flags and sub-projects can override by appending to default flags (and sub-sub-projects can override by appending again).

**Comment by Zoxc** [comment]
s/hierarchical/broken/

File: src/test/run-make-fulldeps/override-aliased-flags/Makefile [hunk 3]
```diff
@@ -0,0 +1,22 @@
+-include ../tools.mk
+
+# FIXME: it would be good to check that it's actually the rightmost flags
+# that are used when multiple flags are specified, but I can't think of a
+# reliable way to check this.
```
Comment (Lonami) [reply src/test/run-make-fulldeps/override-aliased-flags/Makefile]: (cc @varkor for his comment on https://github.com/rust-lang/rust/pull/60426#discussion_r279932992 which I read just now)

Comment (Lonami) [inline src/test/run-make-fulldeps/override-aliased-flags/Makefile]: Can't we make `rustc` show or warn to `stderr` about which flag it chose? "Note: both foo and bar specified. Using bar", then, we could just capture its output and compare.

**Pull request merged**